### PR TITLE
Fix issue #303

### PR DESCRIPTION
**Pull request by jojo1981**
Fixed: Nullable return method "must be an instance of {class name}, null returned"

### CODE REVIEW

File: fixtures/WithReturnTypehints70.php [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 namespace Fixtures\Prophecy;
 
-class WithReturnTypehints extends EmptyClass
+class WithReturnTypehints70 extends EmptyClass
```
Comment (stof) [inline fixtures/WithReturnTypehints70.php]: why adding 70 here ? This class is not specific to PHP 7.0

File: fixtures/WithReturnTypehints70.php [hunk 2]
```diff
@@ -15,4 +15,8 @@ public function getName(): string {
     public function getParent(): parent {
         return $this;
     }
+
+    public function getNullableReturnType() : \stdClass {
```
Comment (stof) [inline fixtures/WithReturnTypehints70.php]: this is *not* a nullable return type, so this is a lie

File: fixtures/WithReturnTypehints71.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Fixtures\Prophecy;
+
+class WithReturnTypehints71 extends EmptyClass
```
Comment (stof) [inline fixtures/WithReturnTypehints71.php]: You should instead have a ``WithNullableReturnTypes`` class containing only methods with nullable types, and used by **different** tests that the existing class (and these tests being skipped before PHP 7.1)

File: tests/Doubler/Generator/ClassMirrorTest.php [hunk 4]
```diff
@@ -335,20 +335,48 @@ public function it_does_not_throw_exception_for_virtually_private_finals()
      */
     public function it_reflects_return_typehints()
     {
-        $class = new \ReflectionClass('Fixtures\Prophecy\WithReturnTypehints');
+        if (version_compare(PHP_VERSION, '7.1', '>=')) {
+            $class = new \ReflectionClass('Fixtures\Prophecy\WithReturnTypehints71');
+        } else {
+            $class = new \ReflectionClass('Fixtures\Prophecy\WithReturnTypehints70');
```
Comment (stof) [inline tests/Doubler/Generator/ClassMirrorTest.php]: this is a no-go

File: tests/Doubler/Generator/ClassMirrorTest.php [hunk 5]
```diff
@@ -335,20 +335,48 @@ public function it_does_not_throw_exception_for_virtually_private_finals()
...
+        $this->assertFalse($classNode->getMethod('getSelf')->isReturnTypeNullable());
+
         $this->assertEquals('\Fixtures\Prophecy\EmptyClass', $classNode->getMethod('getParent')->getReturnType());
+        $this->assertFalse($classNode->getMethod('getParent')->isReturnTypeNullable());
+
+        $this->assertEquals('\stdClass', $classNode->getMethod('getNullableReturnType')->getReturnType());
+        if (version_compare(PHP_VERSION, '7.1', '>=')) {
+            $this->assertTrue($classNode->getMethod('getNullableReturnType')->isReturnTypeNullable());
+        } else {
+            $this->assertFalse($classNode->getMethod('getNullableReturnType')->isReturnTypeNullable());
```
Comment (stof) [inline tests/Doubler/Generator/ClassMirrorTest.php]: this does not make any sense

File: fixtures/WithReturnTypehints70.php (hunk 1 above)
Comment (jojo1981) [inline fixtures/WithReturnTypehints70.php]: I agree this is not OK I will fix it.

File: fixtures/WithReturnTypehints70.php (hunk 2 above)
Comment (jojo1981) [inline fixtures/WithReturnTypehints70.php]: Agree, also this I will fix.

File: fixtures/WithReturnTypehints71.php (hunk 3 above)
Comment (jojo1981) [inline fixtures/WithReturnTypehints71.php]: I agree this is a far better solution. So I will fix it.

File: tests/Doubler/Generator/ClassMirrorTest.php (hunk 5 above)
Comment (jojo1981) [inline tests/Doubler/Generator/ClassMirrorTest.php]: Agree I will fix it and have only a PHP 7.1 check.

File: tests/Doubler/Generator/ClassMirrorTest.php (hunk 4 above)
Comment (jojo1981) [inline tests/Doubler/Generator/ClassMirrorTest.php]: I will fix it.

Comment (unfunco) [inline tests/Doubler/Generator/ClassMirrorTest.php]: `PHP_VERSION_ID` is available as of 5.2.7 too, and Prophecy supports 5.3+ so it's safe to use, and will have slightly better performance when compared with `version_compare`.

File: fixtures/WithReturnTypehints71.php (hunk 3 above)
Comment (jojo1981) [inline fixtures/WithReturnTypehints71.php]: Agree. Added a new test and make use of the PHP Unit `@requires` annotation.

File: tests/Doubler/Generator/ClassMirrorTest.php (hunk 4 above)
Comment (jojo1981) [inline tests/Doubler/Generator/ClassMirrorTest.php]: I have reverted this and added a new nullable return types test.

**Comment by jojo1981** [comment]
@stof thanks for the review. I have processed all the comments and updated this PR. Before pushing I had squashed the commits.

**Comment by jojo1981** [comment]
@stof can you please check this PR again?

**Comment by fgheorghe** [comment]
Would be great to have this reviewed as it is blocking proper testing of nullable methods.

**Comment by Sam-Burns** [comment]
Other than the need fix the `composer.json` conflict, it seems like this PR is mergeable now.  It looks like all the feedback from the code review has been addressed, and it would be good to have working support for nullable types.

**Comment by everzet** [comment]
Please rebase.

**Comment by jojo1981** [comment]
This PR has no purpose anymore because this PR is merged: https://github.com/phpspec/prophecy/pull/322

**Pull request closed without merging**